IndependentMOGP with inputs grouped by feature

willtebbutt opened the pull request:
Currently the `IndependentMOGP` only has support for input vectors whose type is `MOInputIsotopicByOutputs`. This PR adds support for `MOInputIsotopicByFeatures`.

Something annoying happened with the things in the test file, which meant that the diff looks unpleasant. Not sure how to fix that. Apologies! Nothing in the `@testset "MOInputIsotopicByOutputs" begin` block has changed, just the `@testset "MOInputIsotopicByFeatures" begin` block has been added.

Hopefully the tests are sufficient to convince you of correctness -- I've constructed a `GP` equivalent to an `IndependentMOGP` using the `LinearMixingModelKernel` from `KernelFunctions` with identity mixing matrix, and checked that they agree on the various quantities of interest.

thomasgudjonwright commented on src/independent_mogp.jl:
```diff
@@ -115,10 +115,79 @@ Posterior implementation for isotopic inputs, given diagonal Σy (OILMM).
+    <:IndependentMOGP,<:MOInputIsotopicByFeatures,<:Diagonal{<:Real,<:Fill}
+}
+
+function reorder_features_to_outputs_indices(x::MOInputIsotopicByOutputs)
```
This can probably happen after, but should functionality that converts `MOInputIsotopicByOutputs` to `MOInputIsotopicByFeatures` and vice-versa live in `KernelFunctions`?

willtebbutt replied:
Yeah, that would probably be good. I'll open an issue.

thomasgudjonwright replied:
👍🏻 , resolving then!

thomasgudjonwright commented on src/independent_mogp.jl:
```diff
@@ -115,10 +115,79 @@ Posterior implementation for isotopic inputs, given diagonal Σy (OILMM).
+    <:IndependentMOGP,<:MOInputIsotopicByFeatures,<:Diagonal{<:Real,<:Fill}
+}
+
+function reorder_features_to_outputs_indices(x::MOInputIsotopicByOutputs)
```
Maybe I don't fully understand your naming here but should this be changed to this:
```suggestion
function reorder_outputs_to_features_indices(x::MOInputIsotopicByOutputs)
```

thomasgudjonwright replied:
Or even just two `reorder_indices` dispatches?

willtebbutt replied:
The message I'm trying to convey is "this function produces indices which are used to reorder things ordered by features into things ordered by outputs" -- I don't _think_ that they can also be used to go the other way, so I think it's probably important to make the distinction.

With that in mind, what name do you think might be better? I agree that the current one is a bit ambiguous.

I should probably also add a comment saying this...

thomasgudjonwright replied:
Hmm I need to see what this is actually doing then, since it must be different from what I was thinking it was doing. My initial confusion was how do we reorder something by features to by outputs if it is already a `MOInputIsotopicByOutputs`, but this must be me making assumptions as to what the function is doing

thomasgudjonwright replied:
From my understanding and after playing with the functions, `reorder_outputs_to_features_indices(x::MOInputIsotopicByOutputs)` returns the indices that would order it as a `MOInputIsotopicByFeatures` and `reorder_outputs_to_features_indices(x::MOInputIsotopicByFeatures)` returns the indices that would order it as a `MOInputIsotopicByOutputs`. Is that not the case?

willtebbutt replied:
@thomasgudjonwright sorry for the amazing delay here. I've finally gotten around to revisiting this. I think you're right, and have left comments to that effect. Does this resolve this?

thomasgudjonwright replied:
Equally sorry for the delay! The to-do pile can get large pushing some things back eh 😄 These comments definitely clarify it for me. I guess I still find it slightly counterintuitive that something that **reorders by-output indices to by-feature indices** would be a function called `reorder_features_to_outputs_indices` since it sounds like its suggest a change from features to outputs. Not sure if that makes any sense

thomasgudjonwright replied:
Resolving given this is fairly outdated and has been addressed elsewhere

thomasgudjonwright commented on test/independent_mogp.jl:
```diff
@@ -1,78 +1,100 @@
+            test_internal_abstractgps_interface(rng, f, x_train_mo, x_test_mo)
         end
     end
+    @testset "MOInputIsotopicByFeatures" begin
```
I don't know the answer to this, but here do we want to compare LMMs made using `MOInputIsotopicByFeatures` to the naive implementation, or to LMMs made using `MOInputIsotopicByOutputs` for correctness?

Maybe it doesn't make a difference given we test LMMs built with `MOInputIsotopicByOutputs`s against the naive so the two are equivalent really at this point.

willtebbutt replied:
Yeah, I think testing both against a naive implementation should be fine (by transitivity).

thomasgudjonwright reviewed (commented):
Left some comments! Thanks so much for getting this done :)

thomasgudjonwright commented on test/independent_mogp.jl:
```diff
@@ -95,6 +95,11 @@
+
+        # Mix of by-features and by-outputs should also work efficiently.
+        y = MOInputIsotopicByOutputs(collect(range(0.0, 3.0; length=4)), 2)
+        test_internal_abstractgps_interface(rng, f, x, y)
```
Are there any additional tests we want to run to ensure the correctness of `cov` in these cases?

willtebbutt replied:
Yes. I'll check against naive computation. Good catch.

thomasgudjonwright commented on src/independent_mogp.jl:
```diff
@@ -129,10 +129,12 @@ const IsotropicByFeaturesFiniteIndependentMOGP = FiniteGP{
     return sortperm(vec(reshape(1:length(x), x.out_dim, length(x.x))'))
 end
 
+# Indices which, when applied to a vector ordered by outputs, will order it by features.
```
```suggestion
# Indices which, when applied to a vector ordered by features, will order it by outputs.
```

thomasgudjonwright commented on src/independent_mogp.jl:
```diff
@@ -129,10 +129,12 @@ const IsotropicByFeaturesFiniteIndependentMOGP = FiniteGP{
     <:IndependentMOGP,<:MOInputIsotopicByFeatures,<:Diagonal{<:Real,<:Fill}
 }
 
+# Indices which, when applied to a vector ordered by features, will order it by outputs.
```
```suggestion
# Indices which, when applied to a vector ordered by outputs, will order it by features.
```

willtebbutt commented:
@thomasgudjonwright I've made another pass over the naming and comment.

I've removed the method which applied to`MOInputIsotopicByFeatures`, because the thing I need is the same in both cases.

I've also renamed the function to make it clearer what it's doing.

Also, why on earth has our CI stopped running...

thomasgudjonwright commented on src/independent_mogp.jl:
```diff
@@ -115,10 +117,98 @@ Posterior implementation for isotopic inputs, given diagonal Σy (OILMM).
+function indices_which_reorder_features_to_outputs(
+    x::Union{MOInputIsotopicByOutputs,MOInputIsotopicByFeatures}
+)
+    return collect(vec(reshape(1:length(x), length(x.x), x.out_dim)'))
```
So if `x` is a `MOInputIsotopicByOutputs` Should `x[indices_which_reorder_features_to_outputs(x)]` return it ordered by features and similarily if `x` is a `MOInputIsotopicByFeaturess` Should `x[indices_which_reorder_features_to_outputs(x)]` return it ordered by outputs? Testing it in the repl it appears to work the one way (ordered by outputs -> ordered by features) but not the other

thomasgudjonwright commented:
> @thomasgudjonwright I've made another pass over the naming and comment.
> 
> I've removed the method which applied to`MOInputIsotopicByFeatures`, because the thing I need is the same in both cases.
> 
> I've also renamed the function to make it clearer what it's doing.
> 
> Also, why on earth has our CI stopped running... 😬



> @thomasgudjonwright I've made another pass over the naming and comment.
> 
> I've removed the method which applied to`MOInputIsotopicByFeatures`, because the thing I need is the same in both cases.
> 
> I've also renamed the function to make it clearer what it's doing.
> 
> Also, why on earth has our CI stopped running... 😬

Seems like the CI is running again 🥳 ! LGTM once the github-actions comments are resolved!

willtebbutt commented:
> Seems like the CI is running again 🥳 ! LGTM once the github-actions comments are resolved!

On my screen it appears that only the format suggestions are running :( None of the tests appear to be run. Definitely not keen to merge without them running.

willtebbutt commented:
Ah. The repository had apparently disabled CI due to inactivity. Seems to be running fine now. Will merge when CI passes.
